website/registrations/utils.py

```diff
@@ -0,0 +1,349 @@
+        self.schema_questions = BulkRegistrationUpload.get_questions_validations(self.registration_schema)
+        self.validations = {**self.schema_questions, **METADATA_FIELDS}
+        self.errors = []
+        self.rows = [Row(row,
+                         self.validations,
+                         functools.partial(self.log_error, row_index=self.reader.line_num))
+                     for row in self.reader]
+
+    @classmethod
+    def get_questions_validations(cls, registration_schema):
+        schema_questions = {}
+        get_question_validations = lambda question: {'required': question.get('required', False),
+                                   'format': question.get('format', ''),
+                                   'type': question.get('type', 'string'),
+                                   'options': question.get('options', [])}
+        for page in registration_schema.schema['pages']:
+            for question in page['questions']:
+                schema_questions[question['qid']] = get_question_validations(question)
+                if 'properties' in question:
+                    for nested_question in question['properties']:
+                        qid = '{}{}'.format(question['qid'], nested_question['id'])
+                        schema_questions[qid] = get_question_validations(nested_question)
+        return schema_questions
+
+    def log_error(self, **kwargs):
+        self.errors.append({
+            'header': kwargs['header'],
+            # TODO: add a method to generate human friendly excel column index
+            'column_index': kwargs['column_index'],
+            'row_index': kwargs['row_index'],
+            'missing': kwargs.get('missing', False),
+            'invalid': kwargs.get('invalid', False)
+        })
+
+    def validate_csv_header_list(self):
+        expected_headers = self.validations.keys()
+        actual_headers = self.headers
+        diff = set(expected_headers) - set(actual_headers)
+        if len(diff):
+            raise ValidationError('Invalid csv headers: {}'.format(','.join(diff)))
+
+    def get_parsed(self):
+        parsed = []
+        for row in self.rows:
+            parsed.append({'csv_raw': row.get_raw_value(), 'csv_parsed': row.get_parsed_value()})
+        return {'schema_id': self.schema_id, 'registrations': parsed}
+
+    def validate(self):
+        self.validate_csv_header_list()
+        for row in self.rows:
+            row.validate()
+
+class Row():
+    def __init__(self, row_dict, validations, log_error):
+        self.row_dict = row_dict
+        self.cells = [Cell(header.lower(),
+                           value, validations[header.lower()],
+                           functools.partial(log_error, column_index=column_index))
+                           for column_index, (header, value) in enumerate(row_dict.items())]
+
+    def to_json(self):
```

Comment:
is the `to_json` method here useful? It seems we will be passing parsed contents around in python dicts, and I don't see the `to_json` method invoked elsewhere in the classes.

Reply:
this method can go, I added before we changed the expected input for the registration create task